gnucash/gnome-utils/gnc-account-sel.c

```diff
@@ -136,151 +227,453 @@ gnc_account_sel_class_init (GNCAccountSelClass *klass)
+                      G_CALLBACK(toggle_hidden_cb), gas);
+
+    gtk_menu_popup_at_pointer (GTK_MENU(menu), (GdkEvent *)event);
+}
 
-    atnd.gas        = gas;
-    atnd.outList    = NULL;
+static gboolean
+account_is_included (GNCAccountSel *gas, Account *acc)
+{
+    gboolean included = TRUE;
 
-    g_list_foreach (accts, gas_filter_accounts, (gpointer)&atnd);
-    g_list_free (accts);
-    atnd.outList = g_list_reverse (atnd.outList);
+    if (gas->acctExcludeList)
+    {
+        if (g_list_find (gas->acctExcludeList, acc) != NULL)
+            included = FALSE;
+    }
 
-    gtk_list_store_clear (gas->store);
-    for (ptr = atnd.outList, i = 0; ptr; ptr = g_list_next(ptr), i++)
+    /* Filter as we've been configured to do. */
+    if (gas->acctTypeFilters)
     {
-        acc = ptr->data;
-        name = gnc_account_get_full_name (acc);
-        gtk_list_store_append (gas->store, &iter);
-        gtk_list_store_set (gas->store, &iter,
-                            ACCT_COL_NAME, name,
-                            ACCT_COL_PTR,  acc,
-                            -1);
-        if (g_utf8_collate (name, currentSel) == 0)
-            active = i;
-        g_free (name);
+        /* g_list_find is the poor-mans '(member ...)', especially
+         * easy when the data pointers in the list are just casted
+         * account type identifiers. */
+        if (g_list_find (gas->acctTypeFilters,
+                         GINT_TO_POINTER(xaccAccountGetType (acc))) == NULL)
+        {
+            included = FALSE;
+        }
     }
 
-    /* If the account which was in the text box before still exists, then
-     * reset to it. */
-    if (active != -1)
-        gtk_combo_box_set_active (GTK_COMBO_BOX(gas->combo), active);
+    if (gas->acctCommodityFilters)
+    {
+        if (g_list_find_custom (gas->acctCommodityFilters,
+                                GINT_TO_POINTER(xaccAccountGetCommodity (acc)),
+                                gnc_commodity_compare_void) == NULL)
+        {
+            included = FALSE;
+        }
+    }
+    return included;
+}
 
-    g_signal_handlers_unblock_by_func (gas->combo, combo_changed_cb , gas);
+static gboolean
+visible_func (GtkTreeModel *model, GtkTreeIter *iter, gpointer user_data)
+{
+    GNCAccountSel *gas = GNC_ACCOUNT_SEL(user_data);
+    Account *acc;
+    gboolean visible = TRUE;
 
-    g_list_free (atnd.outList);
-    if (currentSel)
-        g_free (currentSel);
+    gtk_tree_model_get (GTK_TREE_MODEL(gas->store), iter, ACCT_COL_PTR, &acc, -1);
+
+    if (acc)
+    {
+        visible = account_is_included (gas, acc);
+
+        if (gas->hide_placeholder && xaccAccountGetPlaceholder (acc))
+            visible = FALSE;
+
+        if (gas->hide_placeholder && xaccAccountIsHidden (acc))
+            visible = FALSE;
+    }
+    return visible;
 }
 
 static void
-gas_filter_accounts (gpointer data, gpointer user_data)
+row_deleted_in_store_cb (GtkTreeModel *model, GtkTreePath *path, gpointer user_data)
```

Comment:
`row_deleted` sounds like a question. `maybe_delete_row_in_store_cb` or `ensure_row_deleted_in_store` would be better.

Reply:
Changed.